### PR TITLE
Fixes #11618 - Replace validation tests by shoulda-matchers

### PR DESCRIPTION
**Pull request by dLobatog**
A good chunk of our unit tests are testing whether a validation is
working or not by testing it actively. For the validations we've
added ourselves I would say it's fine. However for validations that
 come from the Rails framework, we're essentially testing their job.

Instead of testing (for instance) validates(:name, :uniqueness =>
true) by creating two objects and verifying the second one won't
save (these tests are already done at the framework level), we
should simply test we're including validations in our models.

The well-maintained gem shoulda-matchers provide easy functions to
check we're including such validations, and other helpers we can
use to refactor our tests further and not test Rails functionality
twice.

In addition to that we should get a good speed after we refactor the whole suite to this. I've quickly benchmarked this against current develop by running tests 5 times in both branches, and the speedup seems to be around 5 to 10s (not the most rigorous experiment :smile: ).

This PR is just meant to test whether this would be something that would be accepted. If so feel free to merge and I can follow up this one by using 'refs' and fixing more validations later on.

### CODE REVIEW

**Comment by unorthodoxgeek** [comment]
does this syntax work with rails 4? (I believe it shoulda)
if so, ok for merge :+1:

File: bundler.d/test.rb [hunk 1]
```diff
@@ -13,4 +13,5 @@
   gem 'factory_girl_rails', '~> 4.5', :require => false
   gem 'rubocop-checkstyle_formatter', '~> 0.1'
   gem "poltergeist"
+  gem 'shoulda-matchers'
```
Comment (domcleal) [inline bundler.d/test.rb]: This should be pinned to a major version, as v3 will require an additional configuration step and other breakages: https://robots.thoughtbot.com/shoulda-matchers-3-0

File: test/unit/auth_sources/auth_source_ldap_test.rb [hunk 2]
```diff
@@ -6,33 +6,12 @@ def setup
     User.current = users(:admin)
   end
 
-  test "should exists a name" do
```
Comment (domcleal) [inline test/unit/auth_sources/auth_source_ldap_test.rb:10]: I can't see a replacement for this test, only one that checks the max length.

File: test/unit/realm_test.rb [hunk 3]
```diff
@@ -8,42 +8,30 @@ def setup
...
-    host = FactoryGirl.create(:host, :realm => @realm)
-
-    assert host.save
-
-    realm = host.realm
-    assert !realm.destroy
-    assert_match /is used by/, realm.errors.full_messages.join("\n")
+    assert have_many(:hosts).
+      class_name('Host::Managed').
+      dependent(:destroy)
```
Comment (domcleal) [inline test/unit/realm_test.rb]: This looks wrong, why is it passing?  Realms should _not_ be destroyed when hosts are removed, or vice versa.

File: test/unit/realm_test.rb [hunk 4]
```diff
@@ -8,42 +8,30 @@ def setup
...
 
   test "should not destroy if it contains hosts" do
-    disable_orchestration
-    host = FactoryGirl.create(:host, :realm => @realm)
-
-    assert host.save
-
-    realm = host.realm
-    assert !realm.destroy
-    assert_match /is used by/, realm.errors.full_messages.join("\n")
```
Comment (domcleal) [inline test/unit/realm_test.rb:32]: The replacement isn't testing the same thing - this is testing the before_destroy with EnsureNotUsedBy.

File: test/unit/realm_test.rb [hunk 5]
```diff
@@ -8,42 +8,30 @@ def setup
...
-    location2 = Location.create :name => "Switzerland"
-    assert location2.save!
-
-    realm = Realm.create :name => "test.net", :realm_proxy => smart_proxies(:realm), :realm_type => "FreeIPA"
-    realm.locations.destroy_all
-    realm.locations.push location1
-    realm.locations.push location2
-    assert realm.save!
+    assert have_many(:locations).
+      through(:taxable_taxonomies).
```
Comment (domcleal) [inline test/unit/realm_test.rb]: I'd suggest putting a shortcut into test_helper for this, as it'll probably be useful elsewhere.

File: test/unit/realm_test.rb [hunk 6]
```diff
@@ -8,42 +8,30 @@ def setup
...
-    assert location1.save!
-
-    location2 = Location.create :name => "Switzerland"
-    assert location2.save!
-
-    realm = Realm.create :name => "test.net", :realm_proxy => smart_proxies(:realm), :realm_type => "FreeIPA"
-    realm.locations.destroy_all
-    realm.locations.push location1
-    realm.locations.push location2
-    assert realm.save!
```
Comment (domcleal) [inline test/unit/realm_test.rb:46]: I'm unsure if Taxonomix is sufficiently tested independently of standard models to remove this entirely.

File: test/unit/user_test.rb [hunk 7]
```diff
@@ -34,23 +32,11 @@ def setup
   end
 
   test "login should be unique" do
-    u = User.new :auth_source => auth_sources(:one), :login => "foo", :mail  => "foo@bar.com"
-    refute_valid u, :login
-  end
-
-  test "login should also be unique across usergroups" do
```
Comment (domcleal) [inline test/unit/user_test.rb:41]: This isn't tested, it's a custom validator: name_used_in_a_usergroup

File: test/unit/user_test.rb [hunk 8]
```diff
@@ -67,51 +53,37 @@ def setup
...
-    refute @user.save
-
-    @user.firstname = "C_r'a-z.y( )<,Na=me;>"
-    assert @user.save
-
-    @user.firstname = "é ô à"
-    assert @user.save
+    assert allow_value('é ô à').for(:firstname)
+    assert allow_value("C_r'a-z.y( )<,Na=me;>").for(:firstname)
+    @user.firstname = 'The Riddle?'
```
Comment (domcleal) [inline test/unit/user_test.rb]: Can you use refute + allow_value?

File: bundler.d/test.rb (hunk 1 above)
Comment (domcleal) [inline bundler.d/test.rb]: I'd suggest either adopting 3.0.0-rc1 now or waiting until 3.0 final.. it might be easier than adopting 2.x, only to have to immediately change it.

**Comment by domcleal** [comment]
Fine by me, I think it's a good improvement and removes much of the inconsistencies and probably bugs present in our tests.  It'll make me happier to insist on more tests on our validations knowing that the cost is much lower.

File: test/functional/basic_rest_response_test.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+module BasicRestResponseTest
```
Comment (domcleal) [inline test/functional/basic_rest_response_test.rb]: This looks unused?  There are two similar files.

File: test/functional/auth_source_ldaps_controller_test.rb [hunk 10]
```diff
@@ -1,15 +1,11 @@
...
-  def test_index
-    get :index, {}, set_session_user
-    assert_template 'index'
-  end
-
-  def test_new
-    get :new, {}, set_session_user
-    assert_template 'new'
-  end
+  include ::BasicRestResponseTest
```
Comment (domcleal) [inline test/functional/auth_source_ldaps_controller_test.rb]: It might be easier to include this straight into the TestCase class in test_helper, or at least require the file from test_helper as it's used so much.

File: test/functional/environments_controller_test.rb [hunk 11]
```diff
@@ -1,16 +1,12 @@
 require 'test_helper'
+require 'functional/shared/basic_rest_response_test'
 
 class EnvironmentsControllerTest < ActionController::TestCase
-  test "should get index" do
-    get :index, {}, set_session_user
-    assert_response :success
-    assert_not_nil assigns(:environments)
```
Comment (domcleal) [inline test/functional/environments_controller_test.rb:7]: It's not great that these assertions are removed, they can help check that the index is generating the right content.  (Same for one or two others in single-resource tests.)

File: test/test_helper.rb [hunk 12]
```diff
@@ -1,11 +1,6 @@
 require 'rubygems'
 require 'spork'
 
-require 'simplecov'
```
Comment (domcleal) [inline test/test_helper.rb:4]: Unrelated?

File: config/environments/test.rb [hunk 13]
```diff
@@ -44,8 +44,8 @@
   # Raise exception on mass assignment protection for Active Record models
   config.active_record.mass_assignment_sanitizer = :strict
 
-  #enables a few aliases - context, should, and should_eventually methods
-  config.minitest_spec_rails.mini_shoulda = true
+  # Enables a few aliases - context, should, and should_eventually methods
```
Comment (domcleal) [inline config/environments/test.rb]: Disables?

File: test/unit/user_test.rb [hunk 14]
```diff
@@ -138,7 +93,7 @@ def setup
     end
   end
 
-  context "try to login" do
+  describe "try to login" do
```
Comment (domcleal) [inline test/unit/user_test.rb]: Why change context to describe?  I see you're disabling the minispec rails version, but not sure I get why you're adding shoulda-context to add `context`, aliasing `describe` to `context` and then replacing `context` with `describe`.

File: test/unit/usergroup_test.rb [hunk 15]
```diff
@@ -9,24 +9,10 @@ class UsergroupTest < ActiveSupport::TestCase
...
-
-    refute two.valid?
-  end
-
-  test "name can't be blank" do
-    group = FactoryGirl.build(:usergroup, :name => "")
-    refute group.valid?
-  end
-
-  test "name is unique across user as well as usergroup" do
```
Comment (domcleal) [inline test/unit/usergroup_test.rb:24]: Is there a test for this?

File: test/unit/setting_test.rb [hunk 16]
```diff
@@ -89,11 +94,6 @@ def test_create_exclamation_updates_description
     assert_equal 'Test', s.description
   end
 
-  def test_create_with_missing_attrs_does_not_persist
```
Comment (domcleal) [inline test/unit/setting_test.rb:92]: What replaces this?

File: test/functional/basic_rest_response_test.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+      should render_template('new')
+    end
+
+    context 'GET #edit' do
+      setup { get :edit, { :id => self.class.first }, set_session_user }
+      should respond_with(:success)
+    end
+  end
+end
+
```
Comment (domcleal) [inline test/functional/basic_rest_response_test.rb]: Nitpick, blank line.

File: test/functional/roles_controller_test.rb [hunk 18]
```diff
@@ -50,13 +31,6 @@ class RolesControllerTest < ActionController::TestCase
     assert Role.find_by_name('test role')
   end
 
-  test 'get edit goes to right template' do
-    get :edit, {:id => 1}, set_session_user
-    assert_response :success
-    assert_template 'edit'
-    assert_equal Role.find(1), assigns(:role)
```
Comment (domcleal) [inline test/functional/roles_controller_test.rb:57]: missing this `assigns` check too now

File: test/functional/shared/basic_rest_response_test.rb [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    def basic_edit_test(model)
+      context 'GET #edit' do
+        setup { get :edit, { :id => model }, set_session_user }
+        should respond_with(:success)
+        should render_template(:edit)
+      end
+    end
+  end
+end
+
```
Comment (domcleal) [inline test/functional/shared/basic_rest_response_test.rb]: Nitpick, empty line.

File: test/functional/subnets_controller_test.rb [hunk 20]
```diff
@@ -1,15 +1,10 @@
...
-  def test_index
-    get :index, {}, set_session_user
-    assert_template 'index'
-  end
-
-  def test_new
-    get :new, {}, set_session_user
-    assert_template 'new'
-  end
+  include ::BasicRestResponseTest
```
Comment (domcleal) [inline test/functional/subnets_controller_test.rb]: You can also replace test_edit below.

File: test/unit/auth_sources/auth_source_ldap_test.rb (hunk 2 above)
Comment (domcleal) [inline test/unit/auth_sources/auth_source_ldap_test.rb:10]: Appears to have been fixed at https://github.com/theforeman/foreman/pull/2651/files#diff-6c1e29f55a935fec94f5837557af0c7aR10

File: test/unit/environment_test.rb [hunk 21]
```diff
@@ -8,18 +8,9 @@ def setup
...
-  test "name should be unique" do
-    as_admin do
-      env = Environment.create :name => "foo"
-      env2 = Environment.new :name => env.name
-      refute env2.valid?
-    end
-  end
+  should validate_presence_of(:name)
+  should validate_uniqueness_of(:name)
+  should allow_value('new').for(:name)
```
Comment (domcleal) [inline test/unit/environment_test.rb]: I don't know if this is equivalent.  Does it pick up changes to the friendly_ reserved words?

File: test/unit/parameters/common_parameter_test.rb [hunk 22]
```diff
@@ -0,0 +1,10 @@
+require 'test_helper'
+
+class CommonParameterTest < ActiveSupport::TestCase
+  should validate_presence_of(:name)
```
Comment (domcleal) [inline test/unit/parameters/common_parameter_test.rb:4]: It would be useful to have these tests in all Parameter subclasses, can they go into ParameterValidations or similar?

File: test/unit/parameters/domain_parameter_test.rb [hunk 23]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+require 'unit/parameters/parameter_validations'
+
+class DomainParameterTest < ActiveSupport::TestCase
+  include ::ParameterValidations
+end
+
```
Comment (domcleal) [inline test/unit/parameters/domain_parameter_test.rb]: Nitpick, blank line, ditto below.

File: test/unit/parameters/common_parameter_test.rb (hunk 22 above)
Comment (domcleal) [inline test/unit/parameters/common_parameter_test.rb:4]: Sorry, I specifically meant the allow_value type ones below - as the name's already in PV.

File: test/unit/realm_test.rb (hunk 6 above)
Comment (domcleal) [inline test/unit/realm_test.rb:46]: Unaddressed comment?

File: test/unit/parameters/parameter_validations.rb [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+    should belong_to(parameter_class).with_foreign_key(:reference_id)
+  end
+
+  module ClassMethods
+    def parameter_class
+      class_name.split('ParameterTest')[0].downcase.to_sym
+    end
+
+    def class_name
+      return 'HostgroupParameterTest' if self.name == 'GroupParameterTest'
```
Comment (domcleal) [inline test/unit/parameters/parameter_validations.rb]: I'd suggest you just define the `parameter_class` method in each test.

File: test/unit/setting_test.rb (hunk 16 above)
Comment (dLobatog) [inline test/unit/setting_test.rb:92]: validate_presence_of(:name)

Comment (dLobatog) [inline test/unit/setting_test.rb:92]: Nevermind, I'll put it back

File: test/unit/realm_test.rb (hunk 6 above)
Comment (dLobatog) [inline test/unit/realm_test.rb:46]: I can put it back, but I would think this is what https://github.com/theforeman/foreman/pull/2651/files#diff-d07b2223fd71cfc129796bad59f95ebdR12 is doing

File: test/unit/environment_test.rb (hunk 21 above)
Comment (dLobatog) [inline test/unit/environment_test.rb]: What reserved words?

Comment (domcleal) [inline test/unit/environment_test.rb]: The environment model has "new" _removed_ from its list of reserved words in the friendly_id call at the top of the model.  The removed test below about the name "new" is specifically to check that friendly_id doesn't prevent the environment being created.  I just wanted to check (sorry, didn't test myself) that this is still tested.

**Comment by dLobatog** [comment]
Fixed all comments (still have to verify what happens with reserved words & environment_test), and now `BasicRestResponseTest` tests assigns both for index and edit.

File: test/unit/environment_test.rb (hunk 21 above)
Comment (domcleal) [inline test/unit/environment_test.rb]: Yeah, this test doesn't detect the 'new' problem when the reserved words are changed, please restore the old test.

**Comment by dLobatog** [comment]
Updated - now it includes a couple more tests on Environment and I restored the one that tests reserved words.

File: test/functional/domains_controller_test.rb [hunk 25]
```diff
@@ -1,15 +1,11 @@
 require 'test_helper'
+require 'functional/shared/basic_rest_response_test'
```
Comment (domcleal) [inline test/functional/domains_controller_test.rb]: Just repeating an earlier unanswered comment:

> It might be easier to include this straight into the TestCase class in test_helper, or at least require the file from test_helper as it's used so much.

File: test/unit/user_test.rb (hunk 7 above)
Comment (dLobatog) [inline test/unit/user_test.rb:41]: Fixed, test was added back

File: test/functional/environments_controller_test.rb (hunk 11 above)
Comment (dLobatog) [inline test/functional/environments_controller_test.rb:7]: Fixed, now it's even better as the shared module tests this in this class & others

File: test/functional/roles_controller_test.rb (hunk 18 above)
Comment (dLobatog) [inline test/functional/roles_controller_test.rb:57]: Fixed as well through basic_edit_test

File: test/functional/subnets_controller_test.rb (hunk 20 above)
Comment (dLobatog) [inline test/functional/subnets_controller_test.rb]: replaced

File: test/unit/parameters/common_parameter_test.rb (hunk 22 above)
Comment (dLobatog) [inline test/unit/parameters/common_parameter_test.rb:4]: All `allow_value` tests are now in the shared tests class

File: test/functional/domains_controller_test.rb (hunk 25 above)
Comment (dLobatog) [inline test/functional/domains_controller_test.rb]: I'll do that, but I don't really see the point to have the module included in classes that don't use any of the methods

Comment (domcleal) [inline test/functional/domains_controller_test.rb]: Yeah, that's OK by me I think.  Since it's always going to be required by running any of these test files I think not having to require it in each file will save some lines.

**Comment by dLobatog** [comment]
Latest update includes the `BasicRestResponseTest` module directly on every functional test class through the test_helper.

**Comment by domcleal** [comment]
Would you mind rebasing this on develop, and to convert the new validation test added in 54188a7 to the same style?  Thanks.

**Comment by dLobatog** [comment]
Rebased on develop & converted the whitespace tests to a `should_not allow_value` test.

**Comment by domcleal** [comment]
Merged as d6c08a7, thanks @dLobatog.
